orchestrator/src/graph.ts

```diff
@@ -0,0 +1,128 @@
+import { getOptionalDependencies, getRequiredDependencies } from "./docker";
+export class Graph {
+    nodes: Map<string, GraphNode>;
+  
+    constructor() {
+      this.nodes = new Map();
+    }
+  
+    addNode(preprocessors: (string | number)[]): GraphNode {
+      const name = preprocessors[0] as string; 
+      if (!this.nodes.has(name)) {
+        const newNode = new GraphNode(preprocessors);
+        this.nodes.set(name, newNode);
+      }
+      return this.nodes.get(name)!;
+    }
+
+    async constructGraph(P: (string | number)[][], H: (string | number)[][]) : Promise<Set<GraphNode>>  {
+      const R = new Set<GraphNode>();
+      const combinedArray = P.concat(H);
+
+      const Pset = new Set<string>();
+      for(const preprocessor of P){
+        Pset.add(preprocessor[0] as string);
+      }
+      
+      for (const preprocessor of combinedArray) {
+        const requiredPreprocessors = await getRequiredDependencies(preprocessor[0] as string, P);
+        const optionalPreprocessors = await getOptionalDependencies(preprocessor[0] as string, P);
+
+        if(optionalPreprocessors && requiredPreprocessors && requiredPreprocessors.every((r) => Pset.has(r[0] as string))){
+          const node = this.addNode(preprocessor);
+          const tmp = optionalPreprocessors.filter((o) => Array.isArray(o) && Pset.has(o[0] as string));
+
+          if (requiredPreprocessors.length === 0 && tmp.length === 0) {
+            R.add(node);
+          } else {
+            for(const reqPrep of requiredPreprocessors){
+              const nodePrime = this.addNode(reqPrep);
+              node.parents.add(nodePrime);
+              nodePrime.children.add(node);
+            }
+
+            for(const optPrep of optionalPreprocessors){
```

Comment:
There may be problems with how `optionalPreprocessors` is being constructed, but this should be `tmp` since we only want to iterate on optional preprocessors that are in the set of available preprocessors.

Reply:
Still think this is relevant.